File: CNN/CNNtoymodel.py
#=======================================================================
#Purpose: Classification of Digits with a CNN
#
#Model:   3 Layer Convolutional Neural Network
#         Trained with Minibatch Gradient Descent
#
#Inputs:  MNIST Database (digits)
#         60,000 training set
#         10,000 test set 
#
#Output:  10 Classes (digits from 1 to 10)
#Version: 10/2016 Roboball (MattK.)
#Link:    https://github.com/roboball/DeepLearningToymodels
#=======================================================================

import numpy as np
import tensorflow as tf
import random
import os, re
import datetime 
import matplotlib.pyplot as plt 
import scipy.io
#plt.close('all') 

########################################################################
#import data from mat-file: MNIST
########################################################################

print('init 3 layer CNN')
print('start loading MNIST data\n')
mat = scipy.io.loadmat('data/MNIST.mat')

#convert to numpy arrays:
imagetrainxx = np.array(mat['imagetrain'])
labeltrain = np.array(mat['labeltrain'],dtype=int)
imagetestxx = np.array(mat['imagetest'])
labeltest = np.array(mat['labeltest'],dtype=int)

#-----------------------------------------------------------------------
#convert labels to one hot encoding
#-----------------------------------------------------------------------
labtrainone = np.zeros((60000, 10))
labtestone = np.zeros((10000, 10))

#convert training labels
for one in range(60000):
	labtrainone[one][labeltrain[one]] = 1

#convert test labels
for one in range(10000):
	labtestone[one][labeltest[one]] = 1

########################################################################
#init, define and train: CNN model
########################################################################

#define parameter-------------------------------------------------------
epochs = 1
learnrate=0.01
train_size = 60000
test_size = 10000
batsize_train=50
batsize_test=2000
batches_train = int(train_size / batsize_train)
batches_test = int(test_size / batsize_test)


#init: cost history
cost_history = np.empty(shape=[0],dtype=float)
#init: accuracy
acc_history = np.empty(shape=[0],dtype=float)
#init: batches correct
correctbatch = np.empty(shape=[0],dtype=float)

#init: grads and vars
gradvar_history = np.empty(shape=[0],dtype=float)
#init: grads and vars
applygrad_history = np.empty(shape=[0],dtype=float)


#init placeholders
x = tf.placeholder(tf.float32, shape=[None, 784])
y_ = tf.placeholder(tf.float32, shape=[None, 10])

#define functions-------------------------------------------------------

#create tf.variables for weights and biases:
def weight_variable(shape):
  initial = tf.truncated_normal(shape, stddev=0.1)
  return tf.Variable(initial)

def bias_variable(shape):
  initial = tf.constant(0.0, shape=shape)
  return tf.Variable(initial)
  
#layer operations:  
def conv2d(x, W):
  return tf.nn.conv2d(x, W, strides=[1, 1, 1, 1], padding='VALID')

def max_pool_2x2(x):
  return tf.nn.max_pool(x, ksize=[1, 2, 2, 1],
                        strides=[1, 2, 2, 1], padding='SAME')
                                           
#init input-------------------------------------------------------------
np.set_printoptions(suppress=True)
np.set_printoptions(precision=15)
np.random.seed(2)

#init: weights/biases---------------------------------------------------
np.random.seed(2)
init2s = np.random.rand(5,5).astype('f') #as float32
init2= init2s.reshape(5,5,1,1)

np.random.seed(3)
init3s = np.random.rand(5,5).astype('f') #as float32
init3= init3s.reshape(5,5,1,1)

np.random.seed(4)
init4 = np.random.rand(4*4,10).astype('f')

#define model parameter-------------------------------------------------
#1.Layer: 1.conv layer                
W_conv1 = tf.Variable(init2)
b_conv1 = bias_variable([1])
#2.Layer: 2.conv layer                  
W_conv2 = tf.Variable(init3)
b_conv2 = bias_variable([1])
#3.Layer: Fc layer
W_fc1 = tf.Variable(init4)
b_fc1 = bias_variable([10])

#list of variables
weightbiaslist = [W_conv1,b_conv1,W_conv2,b_conv2, W_fc1,b_fc1]

# reshape input
x_image = tf.reshape(x, [-1,28,28,1])
#~ x_view =  tf.reshape(x_image, [28,28]) #for view only

########################################################################
#define CNN model
########################################################################

#1.Layer: conv----------------------------------------------------------
h_conv1 = conv2d(x_image, W_conv1) + b_conv1
h_relu1 = tf.nn.relu(h_conv1)
h_pool1 = max_pool_2x2(h_relu1)

#2.Layer: conv----------------------------------------------------------
h_conv2 = conv2d(h_pool1, W_conv2) + b_conv2
h_relu2 = tf.nn.relu(h_conv2)
h_pool2 = max_pool_2x2(h_relu2)

#3.Layer: Densely Connected Layer---------------------------------------
h_flat = tf.reshape(h_pool2, [-1, 4*4*1])
y_conv = tf.matmul(h_flat, W_fc1) + b_fc1

#4.Layer: cross_entropy ------------------------------------------------
softmax = tf.nn.softmax(y_conv)
#Cost function
cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(y_conv, y_))

########################################################################
#backward pass:
########################################################################

#-----------------------------------------------------------------------
#do manual backprop
#-----------------------------------------------------------------------
# Create an optimizer
opt = tf.train.GradientDescentOptimizer(learning_rate = learnrate)
# Compute the gradients for a list of variables.
grads_and_vars = opt.compute_gradients(cross_entropy, weightbiaslist)
# apply the gradients
apply_grads = opt.apply_gradients(grads_and_vars)
#-----------------------------------------------------------------------

#init testing:
#compare labels and predictions
correct_prediction = tf.equal(tf.argmax(y_conv,1), tf.argmax(y_,1))
#tf.cast: converts float64 to float32
#take the mean of the correct_prediction vector
accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))

#init session:----------------------------------------------------------
sess = tf.InteractiveSession()
#sess.run(tf.initialize_all_variables())
sess.run(tf.global_variables_initializer())

########################################################################
#train model:
########################################################################

t1_1 = datetime.datetime.now()
print('=============================================')
print("start CNN training" )
print('=============================================')

for epoch in range(1,epochs+1):
	
	for batch in range(1,batches_train+1):
		
		#grab train batch:
		trainbatchx = imagetrainxx[:,batch-1:batch+batsize_train-1] 
		#reshape the input training batch:
		imt5 = np.empty(shape=[1,784],dtype=float)
		
		for pos2 in range(batsize_train):
			imt = trainbatchx[:,pos2]
			imt2 = np.reshape(imt,(28, 28))
			imt3 = np.transpose(imt2)
			imt4 = np.reshape(imt3,(1, 784))
			imt5 = np.append(imt5,imt4, axis=0)
			#~ plt.figure(pos2+10)
			#~ plt.axis("off")
			#~ imgplot11 = plt.imshow(imt3)
		
		#batch: train images and train labels	
		trainbatch = imt5[1:batsize_train+1,:] 
		trainbatchlab = labtrainone[batch-1:batch+batsize_train-1,:] 
		feedtrain = {x:trainbatch , y_: trainbatchlab}
		
		################################################################
		#printout: input, weights and biases
		################################################################
		
		#~ print(trainbatch.shape)
		#~ print ("inputlabel:\n",trainbatchlab)
		#~ print ("input:")
		
		#show image:
		#~ imttx = np.reshape(trainbatch[2,:],(28, 28))
		#~ plt.figure(25)
		#~ plt.axis("off")
		#~ imgp = plt.imshow(imttx)
		
		#~ print ("Weight1:")
		#~ print (sess.run(W_conv1))
		#~ print ("bias1:")
		#~ print (sess.run(b_conv1))
		#~ print ("Weight2:")
		#~ print (sess.run(W_conv2))
		#~ print ("bias2:")
		#~ print (sess.run(b_conv2))
		#~ print ("Weight3FC:")
		#~ print (sess.run(W_fc1))
		#~ print ("bias3fc:")
		#~ print (sess.run(b_fc1))
		################################################################
		
		################################################################
		#printout: forward pass
		################################################################
		
		#1.Layer--------------------------------------------------------
		#~ print ("1.Layer:--------------------------------")
		#~ print (sess.run(h_conv1,feed_dict= {x: trainbatch}))
		#~ conv1y = sess.run(h_conv1,feed_dict= {x: trainbatch})
		#~ conv1 = tf.reshape(conv1y, [24, 24])
		#~ print ("h_conv1:")
		#~ print (sess.run(conv1))
		#~ relu1y = sess.run(h_relu1,feed_dict= {x: trainbatch})
		#~ relu1 = tf.reshape(relu1y, [24, 24])
		#~ print ("h_relu1:")
		#~ print (sess.run(relu1))
		#~ pool1y = sess.run(h_pool1,feed_dict= {x: trainbatch})
		#~ pool1 = tf.reshape(pool1y, [12, 12])
		#~ print ("h_pool1:")
		#~ print (sess.run(pool1))
		#---------------------------------------------------------------
		
		#2.Layer--------------------------------------------------------
		#~ print ("2.Layer:--------------------------------")
		#~ print (sess.run(h_conv1,feed_dict= {x: trainbatch}))
		#~ conv2y = sess.run(h_conv2,feed_dict= {x: trainbatch})
		#~ conv2 = tf.reshape(conv2y, [8, 8])
		#~ print ("h_conv2:")
		#~ print (sess.run(conv2))
		#~ relu2y = sess.run(h_relu2,feed_dict= {x: trainbatch})
		#~ relu2 = tf.reshape(relu2y, [8, 8])
		#~ print ("h_relu2:")
		#~ print (sess.run(relu2))
		#~ pool2y = sess.run(h_pool2,feed_dict= {x: trainbatch})
		#~ pool2 = tf.reshape(pool2y, [4, 4])
		#~ print ("h_pool2:")
		#~ print (sess.run(pool2))
		#---------------------------------------------------------------
		
		#3.Layer--------------------------------------------------------
		#~ print ("3.Layer:--------------------------------")
		#~ print ("fc matmul check:")
		#~ fctest = sess.run(y_conv,feed_dict= {x: trainbatch})
		#~ print (fctest)
		
		#~ print ("softmax check:")
		#~ softmaxtest = sess.run(softmax,feed_dict= {x: trainbatch})
		#~ print (softmaxtest)
		#---------------------------------------------------------------
		
		################################################################
		#history get backprop values
		################################################################
		##  get the gradients (for update)
		grad_values = sess.run([grad for (grad,var) in grads_and_vars], feed_dict=feedtrain)
		gradvar_history = np.append(gradvar_history,grad_values)
		##  get the weights (var)
		var_values = sess.run([var for (grad,var) in grads_and_vars], feed_dict=feedtrain)
		applygrad_history = np.append(applygrad_history,var_values)
		
		################################################################
		#printout: backward pass
		################################################################
		
		#~ crosscheck = sess.run(cross_entropy,feed_dict=feedtrain)
		#~ print("cross entropy check:\n", crosscheck)
		
		#print gradients:-----------------------------------------------
		#~ print ('grad_values weight1: ')
		#~ print (sess.run(tf.reshape(grad_values[0], [5, 5])))
		#~ print ('grad_values bias1 : ')
		#~ print (grad_values[1])
		#~ print ('grad_values weight2: ')
		#~ print (sess.run(tf.reshape(grad_values[2], [5, 5])))
		#~ print ('grad_values bias2 : ')
		#~ print (grad_values[3])
		#~ print ('grad_values weight3fc: ')
		#~ print (sess.run(tf.reshape(grad_values[4], [4*4, 10])))
		#~ print ('grad_values bias3 : ')
		#~ print (grad_values[5])
		#~ print('=============================================')
		#~ print('=============================================')
		
		####################################################################
		
		if batch%100 == 0:
			train_accuracy = accuracy.eval(feed_dict=feedtrain)
			crossloss = sess.run(cross_entropy,feed_dict=feedtrain)
			t2_1 = datetime.datetime.now()
			print('epoch: '+ str(epoch)+'/'+str(epochs)+
			' -- training batch: '+ str(batch)+'/'+str(batches_train)+
			" -- cross entropy loss: " + str(crossloss))
			print('training accuracy: %.2f'% train_accuracy + 
			" -- training time: " + str(t2_1-t1_1))
		
		##weight update: Minibatch - SGD
		train_step = sess.run(apply_grads, feed_dict=feedtrain)
		
		#get cost_history, accuracy history data:
		cost_history = np.append(cost_history,sess.run(cross_entropy,feed_dict=feedtrain))
		acc_history = np.append(acc_history,sess.run(accuracy,feed_dict=feedtrain))
       
#Measure Training Time:-------------------------------------------------       
t3_1 = datetime.datetime.now()
print('---------------------------------------------')
print("overall training time: " + str(t3_1-t1_1)+'\n')


########################################################################
#testing different datasets:
########################################################################
  
#######################Testing: MNIST data##############################

print('=============================================')
print('start CNN testing')
print('=============================================')

for pos in range(0,test_size,batsize_test):
	testbatchx = imagetestxx[:,pos:pos+batsize_test] #grab test batch
	
	#reshape the input test batch:
	imtx5 = np.empty(shape=[1,784],dtype=float)
	
	for pos2 in range(batsize_test):
		imtx = testbatchx[:,pos2]
		imtx2 = np.reshape(imtx,(28, 28))
		imtx3 = np.transpose(imtx2)
		imtx4 = np.reshape(imtx3,(1, 784))
		imtx5 = np.append(imtx5,imtx4, axis=0)
		#~ plt.figure(pos2+30)
		#~ plt.axis("off")
		#~ imgplot11 = plt.imshow(imtx3)
		
	testbatch = imtx5[1:batsize_test+1,:] #batch: test images	
	testbatchlab = labtestone[pos:pos+batsize_test,:] #batch: test labels
	feedtest = {x: testbatch , y_: testbatchlab}
	
	predictions = accuracy.eval(feed_dict=feedtest)
	correctbatch = np.append(correctbatch,predictions)
	if pos%100 == 0:
		print('test batch: '+str(pos+batsize_test)+'/'+str(test_size)+
		' -- test accuracy: %.2f' % np.mean(correctbatch))
print('---------------------------------------------')
print("overall test accuracy %.3f"%np.mean(correctbatch, axis=0)+'\n') 


########################################################################
#plot settings:
########################################################################

#=======================================================================
#plot training
#=======================================================================
#plot loss function-----------------------------------------------------
plt.figure(1, figsize=(8,8))
plt.subplots_adjust(wspace=0.5,hspace=0.5)
plt.subplot(211)
plt.plot(range(len(cost_history)),cost_history, color='b')
plt.axis([0,batches_train,0,10])
plt.title('Cross Entropy Training Loss')
plt.xlabel('number of batches')
plt.ylabel('loss')
#plot training accuracy function----------------------------------------
plt.subplot(212)
plt.plot(range(len(acc_history)),acc_history, color='g')
plt.axis([0,batches_train,0,1])
plt.title('Training Accuracy')
plt.xlabel('number of batches')
plt.ylabel('accuracy percentage')

#plt.show()
#plt.hold(False)

#=======================================================================
#plot testing
#=======================================================================
#plot loss function-----------------------------------------------------
plt.figure(2, figsize=(8,8))
plt.subplots_adjust(wspace=0.5,hspace=0.5)
plt.subplot(211)
plt.plot(range(len(cost_history)),cost_history, color='b')
plt.axis([0,batches_train,0,np.max(cost_history)])
plt.title('Cross Entropy Training Loss')
plt.xlabel('number of batches')
plt.ylabel('loss')
#plot test accuracy function----------------------------------------
plt.subplot(212)
plt.plot(np.arange(1,batches_test +1, 1.0),correctbatch, color='m')
#plt.axis([int(1),int(numtestbat),0,1])
plt.yticks(np.arange(0,1.2, 0.2))
plt.xticks(np.arange(1,batches_test +1, 1.0))
plt.title('Test Accuracy')
plt.xlabel('number of batches, batch size: '+ str(batsize_test))
plt.ylabel('accuracy percentage')


#plt.ion()
plt.show()
#plt.hold(False)

File: CNN/CNNtoymodel.m
%==========================================================================
%Purpose: Classification of Digits with a CNN
%Model:   3 Layer Convolutional Neural Network
%         Trained with Minibatch Gradient Descent
%Inputs:  MNIST Database (digits)
%         60,000 training set
%         10,000 test set 
%Output:  10 Classes (digits from 1 to 10)
%Version: 10/2016 (roboball MattK.)
%Link: https://github.com/roboball/DeepLearning/blob/master/README.md
%==========================================================================

clear all
close all
clc

%add filepath and sub directories
tmp = matlab.desktop.editor.getActive;
cd(fileparts(tmp.Filename));
addpath(genpath('../CNN'));

%********************************************************
%inputs from mat file: 
%********************************************************
load('data/MNIST.mat')

%********************************************************
%init parameter:
%********************************************************
epochs = 1; %number of iterations
eta = 0.01; %learning rate 
trainsamples = 60000; %60000 training samples
batsize_train = 50; %training batch size 
testsamples = 10000; %10000 test samples
batsize_test = 2000; %test batch size 
prostep = 100; %stepsize for proof of progress
numclass = 10; %number of classes

%number of batches for training and testing:
batches_train = trainsamples / batsize_train; 
batches_test = testsamples / batsize_test; 

%dummy vector: gradient for nnloss
gradloss = ones(1,1,10,1, 'double'); 
%init loss history
totalloss =[];
%init accuracy history
trainacc = [0];
testacc =[];

%********************************************************
%create target vectors: one hot encoding
%********************************************************

%training targets
tar_train = zeros(numclass,length(labeltrain));
for label = 1:length(labeltrain)
        tar_train(labeltrain (label)+1,label) = 1;
end

%training targets
tar_test = zeros(numclass,length(labeltest));
for label = 1:length(labeltest)
        tar_test(labeltest (label)+1,label) = 1;
end

%********************************************************
%init weights:
%********************************************************

%init weights
rng(2); format long
w1 = rand(5,5)';
rng(3); format long
w2 = rand(5,5)';
rng(4); format long
w3 = rand(10,4*4)';

%init biases
b1=0;
b2=0;
b3=zeros(1,10);

%==========================================================================
%start training:
%==========================================================================
disp('================================')
disp('start CNN training')
disp('================================')
tic
for epoch = 1:epochs

for batch = 1:batches_train
  
%get minibatch:
minibatch = imagetrain(:,batch:batch+batsize_train-1);  %grab data
minibatchlabel = tar_train(:,batch:batch+batsize_train-1); %grab labels

for sample = 1:batsize_train
      
%**************************************************************************
% FORWARD PASS
%**************************************************************************

%input from minibatch:
x1 = double(reshape(minibatch(:,sample),[28,28]));
target = minibatchlabel(:,sample);

%1.Layer
l1_conv  = conv2(x1,rot90(w1,2),'valid' ) + b1; 
l1_relu  = relu( l1_conv );
[ l1_pool , l1_storepool ] = maxpool(l1_relu, 2);

%2.Layer
l2_conv  = conv2(l1_pool,rot90(w2,2),'valid') + b2; 
l2_relu  = relu( l2_conv );
[ l2_pool , l2_storepool ] = maxpool(l2_relu, 2);

%3.Layer: FC1
l3_reshape = reshape(l2_pool',1,4*4);
l3_fclayer = w3' * l3_reshape' + b3';

% Output Layer: softmax probs, cross-entropy loss and output deltas
[ softmaxprobs, ce_loss, deltaout ] = softmax( l3_fclayer, target );

%**************************************************************************
% BACKWARD PASS
%**************************************************************************

%********************************************************
%hidden deltas:
%********************************************************
%3.Layer:
l3_backfclayer = deltaout * w3'; 
l3_backreshape = reshape(l3_backfclayer ,4,4)';
%2.Layer:
l2_backpool = maxpoolup( l3_backreshape, l2_storepool , 2 );
l2_backrelu = reluup( l2_backpool, l2_relu );
l2_backconv = conv2(l2_backrelu,w2,'full'); 
%1.Layer:
l1_backpool = maxpoolup( l2_backconv,l1_storepool, 2 );
l1_backrelu = reluup( l1_backpool, l1_relu );
l1_backconv = conv2(l1_backrelu,w1,'full');

%********************************************************
%gradients for update:
%********************************************************
%weight gradients:
grad_weight3 = l3_reshape' * deltaout;
grad_weight2 = conv2(l1_pool,rot90(l2_backrelu,2),'valid');
grad_weight1 = conv2(x1,rot90(l1_backrelu,2),'valid'); 
%bias gradients:
grad_bias3 = deltaout;
grad_bias2 = sum(sum(1 * l2_backrelu)); 
grad_bias1 = sum(sum(1 * l1_backrelu)); 

%********************************************************
%store: gradients and loss
%********************************************************
%weight gradients:
gradw1(:,:,:,sample) = grad_weight1;
gradw2(:,:,:,sample) = grad_weight2;
gradw3(:,:,:,sample) = grad_weight3;
%bias gradients:
gradb1(:,:,:,sample) = grad_bias1;
gradb2(:,:,:,sample) = grad_bias2;
gradb3(:,:,:,sample) = grad_bias3;

%loss
batchloss(sample) = ce_loss;

end

%********************************************************
%gradient update: Minibatch Gradient Descent:
%********************************************************

%average the gradients:
avg_gradw1 = sum(gradw1,4)./batsize_train;
avg_gradw2 = sum(gradw2,4)./batsize_train;
avg_gradw3 = sum(gradw3,4)./batsize_train;
avg_gradb1 = sum(gradb1,4)./batsize_train;
avg_gradb2 = sum(gradb2,4)./batsize_train;
avg_gradb3 = sum(gradb3,4)./batsize_train;

%weight updates:
w1 = w1 - eta * avg_gradw1; 
w2 = w2 - eta * avg_gradw2;
w3 = w3 - eta * avg_gradw3; 
b1 = b1 - eta * avg_gradb1;
b2 = b2 - eta * avg_gradb2;
b3 = b3 - eta * avg_gradb3';

%********************************************************
%average batchloss:
%********************************************************
loss = sum(batchloss)/batsize_train;
epochloss(batch) = loss;

%check training accuracy after every 100 batches
if mod(batch,prostep) == 0
%********************************************************
%training accuracy:
%********************************************************
correct=0;
for sample = 1:batsize_train
      
%**************************************************************************
% FORWARD PASS
%**************************************************************************

%input from minibatch:
x1 = double(reshape(minibatch(:,sample),[28,28]));
target = minibatchlabel(:,sample);

%1.Layer
l1_conv  = conv2(x1,rot90(w1,2),'valid' ) + b1; 
l1_relu  = relu( l1_conv );
[ l1_pool , l1_storepool ] = maxpool(l1_relu, 2);

%2.Layer
l2_conv  = conv2(l1_pool,rot90(w2,2),'valid') + b2; 
l2_relu  = relu( l2_conv );
[ l2_pool , l2_storepool ] = maxpool(l2_relu, 2);

%3.Layer: FC1
l3_reshape = reshape(l2_pool',1,4*4);
l3_fclayer = w3' * l3_reshape' + b3';

% Output Layer: softmax probs, cross-entropy loss and output deltas
[ softmaxprobs1, ce_loss1, deltaout1 ] = softmax( l3_fclayer, target );

%compare softmax prob to target
[maxi_softmax, idx] = max(softmaxprobs1); %get index softmax
idxtarget = find(target == 1); %get index target

if idx == idxtarget
    correct = correct + 1;
end

end

%store: training accuracy
trainacc = [trainacc,correct/batsize_train];

%********************************************************
%show CNN training progress:
%********************************************************    
disp(['epoch ' num2str(epoch) '/' num2str(epochs) ' -- training batch '...
        num2str(batch) '/' num2str(batches_train)...
        ' -- cross entropy loss ' num2str(loss)]); 
disp(['training accuracy: ',num2str(correct/batsize_train)...
    ' -- time: ' num2str(toc/60)])
end

end
totalloss = [totalloss,epochloss];
end
traintime = toc/60;
disp('---------------------------------')
disp(['total training time: ',num2str(traintime),char(10) ])

%==========================================================================
%start testing:
%==========================================================================
disp('================================')
disp('start CNN testing')
disp('================================')

for batch2 = 1:batches_test
  
%get minibatch:
minibatch = imagetest(:,batch2:batch2+batsize_test-1);  %grab batch
minibatchlabel = tar_test(:,batch2:batch2+batsize_test-1); %grab labels

correct=0; %init acc
for sample = 1:batsize_test
      
%**************************************************************************
% FORWARD PASS
%**************************************************************************

%input from minibatch:
x1 = double(reshape(minibatch(:,sample),[28,28]));
target = minibatchlabel(:,sample);

%1.Layer
l1_conv  = conv2(x1,rot90(w1,2),'valid' ) + b1; 
l1_relu  = relu( l1_conv );
[ l1_pool , l1_storepool ] = maxpool(l1_relu, 2);

%2.Layer
l2_conv  = conv2(l1_pool,rot90(w2,2),'valid') + b2; 
l2_relu  = relu( l2_conv );
[ l2_pool , l2_storepool ] = maxpool(l2_relu, 2);

%3.Layer: FC1
l3_reshape = reshape(l2_pool',1,4*4);
l3_fclayer = w3' * l3_reshape' + b3';

% Output Layer: softmax probs, cross-entropy loss and output deltas
[ softmaxprobs2, ce_loss2, deltaout2 ] = softmax( l3_fclayer, target );

%compare softmax prob to target
[maxi_softmax2, idx2] = max(softmaxprobs2); %get index softmax
idxtarget2 = find(target == 1); %get index target

if idx2 == idxtarget2
   correct = correct + 1;
end

end

%store: test accuracy
testacc = [testacc,correct/batsize_test];

%********************************************************
%show CNN testing progress:
%********************************************************    
disp(['test batch ' num2str(batch2) '/' num2str(batches_test) ...
      ' -- test accuracy: ',num2str(correct/batsize_test)]) 
end

disp('---------------------------------')
totaltestacc = sum(testacc)/batches_test;
disp(['total test accuracy: ',num2str(totaltestacc),char(10)])

%**************************************************************************
%plot settings: 
%**************************************************************************

%plot: training error
figure(100)
maxiloss = max(totalloss);
plot(totalloss,'b','LineWidth',2);
%axis([1 batch*epochs+0.001 0 ceil(maxiloss)+1])
axis([1 batch*epochs+0.001 0 10])
title(['Training Error: CNN after ',num2str(epochs),' epochs',...
       ' , batch-size: ',num2str(batsize_train)])
xlabel('batches')
ylabel('total error')

if trainsamples ~= 1

%plot: training accuracy
figure(101)
xax = 0:prostep:batch*epochs;
xai = 0:1:batch*epochs;
vq1 = interp1(xax,trainacc, xai, 'pchip');
%plot(xax,trainacc,'g',xai,vq1,'-g','LineWidth',2 );
plot(xai,vq1,'-g','LineWidth',2 );
axis([0 batch*epochs+0.001 0 1])
title('Training Accuracy ')
xlabel('batches')
ylabel('accuracy in %')

end

%plot: test accuracy
figure(102)
xax2 = 1:batch2;
xai2 = 1:1:batch2;
vq2 = interp1(xax2,testacc, xai2, 'pchip');
%plot(xax2,testacc,'m',xax2,vq2,'-m','LineWidth',2);
plot(xax2,vq2,'-m','LineWidth',2);
axis([1 batch2+0.001 0 1])
title('Test Accuracy ')
xlabel('batches')
ylabel('accuracy in %')

















